### PR TITLE
feat(treesitter)!: consolidate query util functions

### PR DESCRIPTION
**Pull request by lewis6991**
And address more type errors.

### CODE REVIEW

File: runtime/lua/vim/treesitter/query.lua [hunk 1]
```diff
@@ -256,39 +236,48 @@ function M.parse_query(lang, query)
...
+  self.captures = self.info.captures
+  query_cache[lang][query] = self
+  return self
+end
+
+---@param node TSNode
+---@param source integer|string
+---@param metadata TSMetadata
+---@return Range6
+function M.get_range_from_metadata(node, source, metadata)
```
Comment (justinmk) [inline runtime/lua/vim/treesitter/query.lua:253]: Could be named `get_range`. What other kinds of "get range" cases do we anticipate in the future?

**Review by justinmk — APPROVED** [review]

File: runtime/doc/treesitter.txt [hunk 2]
```diff
@@ -789,14 +789,12 @@ get_node_text({node}, {source}, {opts})
       • {source}  (integer|string) Buffer or string from which the {node} is
                   extracted
       • {opts}    (table|nil) Optional parameters.
-                  • concat: (boolean) Concatenate result in a string (default
```
[the remark this answers is not in the document]
Comment (lewis6991) [reply runtime/doc/treesitter.txt:792]: I also need to make sure that using `nvim_buf_get_text` like I have doesn't have any off by one errors. Though those can be easily fixed in follow ups.

[the remark this answers is not in the document]
Comment (clason) [reply runtime/doc/treesitter.txt:792]: (I'll double-check, but I believe the `concat = true` code path is actually the one from treesitter, so any errors from removing that option should crop up here -- or in some other plugin.)

File: runtime/doc/treesitter.txt [hunk 3]
```diff
@@ -822,6 +820,18 @@ get_query_files({lang}, {query_name}, {is_included})
         string[] query_files List of files to load for given query and
         language
 
+get_range({node}, {source}, {metadata})           *vim.treesitter.get_range()*
+    Get the range of a |TSNode|. Can also supply {source} and {metadata} to
```
Comment (clason) [inline runtime/doc/treesitter.txt:824]: what is `{source}`, conceptually? `{metadata}` is pretty clear from context, but `{source}` is very generic and not a common term in these docs (yet)

File: runtime/lua/vim/treesitter/query.lua [hunk 4]
```diff
@@ -256,39 +236,51 @@ function M.parse_query(lang, query)
...
 --- Gets the text corresponding to a given node
 ---
 ---@param node TSNode
 ---@param source (integer|string) Buffer or string from which the {node} is extracted
 ---@param opts (table|nil) Optional parameters.
----          - concat: (boolean) Concatenate result in a string (default true)
 ---          - metadata (table) Metadata of a specific capture. This would be
 ---            set to `metadata[capture_id]` when using |vim.treesitter.add_directive()|.
----@return (string[]|string|nil)
+---@return string
```
Comment (mfussenegger) [inline runtime/lua/vim/treesitter/query.lua:270]: Very nice,  that this returned different types depending on the options and required casts or asserts to silence luals bothered me just yesterday 👍

File: test/functional/treesitter/parser_spec.lua [hunk 5]
```diff
@@ -277,28 +277,6 @@ void ui_refresh(void)
     eq('void', res2)
   end)
 
-  it('support getting text where start of node is past EOF', function()
```
Comment (lewis6991) [inline test/functional/treesitter/parser_spec.lua:280]: @mfussenegger is this still something you think we should support. My intuition says if you pass an invalid node, you get an error.

Comment (mfussenegger) [reply test/functional/treesitter/parser_spec.lua:280]: No strong opinion on it, I added this based on https://github.com/neovim/neovim/pull/14999#discussion_r666674419

Comment (lewis6991) [reply test/functional/treesitter/parser_spec.lua:280]: Ah ok. So basically we just need to cover the case when a range is zero width at EOF. That should be covered ok now whilst being stricter than before.

File: runtime/lua/vim/treesitter/query.lua (hunk 4 above)
Comment (lewis6991) [reply runtime/lua/vim/treesitter/query.lua:278]: Yeah, this has been bothering me a lot.

File: runtime/doc/news.txt [hunk 6]
```diff
@@ -200,6 +200,11 @@ The following new APIs or features were added.
   https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection .
   Support for the previous format will be removed in a future release.
 
+• Changes to |vim.treesitter.get_node_text()|:
```
Comment (clason) [inline runtime/doc/news.txt:203]: should be under "Changed features" below, no?

Comment (clason) [reply runtime/doc/news.txt:203]: Same for the above item, really.

Comment (lewis6991) [reply runtime/doc/news.txt:203]: I'd say breaking changes. Wrong section either way.

Comment (clason) [reply runtime/doc/news.txt:203]: (actually, it looks like people just dumped everything under "new" (i.e., newly _added_) features. will have to clean up before release...)

**Review by clason — APPROVED** [review]
News is not a blocker.

File: runtime/doc/news.txt (hunk 6 above)
Comment (clason) [reply runtime/doc/news.txt]: That works, too :D

**Pull request merged**